[Dawn] Initial Implementation of Notifications; Faster Rendering Rates

kevinmasd opened the pull request:
Initial Implementation of Notifications; untested with real robot so far.

Info-Actions and Ansible modified to reduce state dispatch calls to reduce rendering rates. Editor still slow, and future speed changes will come in future improvements.

kevinmasd committed:
[Dawn] IP Changes were not propagated to main

kevinmasd committed:
[Dawn] First Implementation of ESTOP, Notification TCP Protocol

kevinmasd committed:
[Dawn] Working on NameEdit

Commit:
changed gamePad prop type to String

kevinmasd committed:
[Dawn] Test Counter-Based Render Throttle

kevinmasd committed:
[Dawn] PeripheralList Cleanup

kevinmasd committed:
[Dawn] FCPL Renamed to PeripheralList

kevinmasd committed:
[Dawn] Reduce dispatch calls per message; peripheral deletion/addition handled in reducer

kevinmasd committed:
[Dawn] Sagas Test modified for messaging changes

nikitakit commented on dawn/.eslintrc:
```diff
@@ -33,6 +33,8 @@
     "import/prefer-default-export": 0,
-    "import/no-extraneous-dependencies": 0
+    "import/no-extraneous-dependencies": 0,
+    "no-empty": 0,
```
Why do you need to disable these linter rules?

kevinmasd replied:
Allow empty loops to act as blocking mechanisms. (See Editor's SCP handling)

kevinji replied:
Wouldn't it make more sense to just disable the rule on a case-by-case basis instead of for the entire project? These rules are kinda useful to keep enabled.

nikitakit commented on dawn/main/ansible/Ansible.js:
```diff
@@ -1,35 +1,103 @@
+    received = true;
+    const decoded = Notification.decode(data);
+    if (decoded.header === Notification.Type.STUDENT_RECEIVED) {
+      RendererBridge.reduxDispatch(notifyChange(0));
```
What is the `0` here?

Would this be better as an enum?

kevinmasd replied:
I'll enum it. It's supposed to match the protobuf numbering.

nikitakit replied:
Oh, in that case is it possible to pull the enum directly out of the `.proto` file, instead of replicating the numbers?

kevinmasd replied:
I'd like to keep a barrier between main/renderer components.

nikitakit replied:
That's fine. Though if it's possible to also import a second copy of the proto file here, then maybe you can do that.

kevinmasd replied:
Actually, thought it was something different. This enum deals with the protocol for notifications (i.e. Dawn -> Runtime Call, Runtime -> Dawn Replay or timeout)

nikitakit replied:
Oh I see

nikitakit commented on dawn/main/ansible/Ansible.js:
```diff
@@ -1,35 +1,103 @@
+    console.log('Runtime Notified');
+  });
+  received = false;
+  RendererBridge.reduxDispatch(notifyChange(1));
```
What does `1` here signify?

nikitakit commented on dawn/main/ansible/Ansible.js:
```diff
@@ -1,35 +1,103 @@
+  }
+  if (!received) {
+    console.error('No Confirmation');
+    RendererBridge.reduxDispatch(notifyChange(2));
```
What does `2` here signify?

nikitakit commented on dawn/renderer/components/PeripheralList.js:
```diff
@@ -1,29 +1,86 @@
 /**
- * A generic component for displaying a list of Peripherals.
+ * Haxors
```
???

kevinmasd replied:
Renamed FCPL to PeripheralList, so this was carried over.

awli replied:
I can't believe FCPL lasted that long...

nikitakit reviewed (commented):
Thanks for fixing the lag!

I tested this with fake runtime, and it can now handle 10Hz (but not 20Hz). My only thought about this change was that it might be good to lower `SENDRATE` in fake runtime to 100 or lower, to make sure that any performance regressions are caught as part of Dawn development.

This PR also contains protocol changes, so I would like a second reviewer to look at that part of the code.

nikitakit commented on dawn/renderer/components/Editor.js:
```diff
@@ -119,6 +120,10 @@ class Editor extends React.Component {
   }
 
+  shouldComponentUpdate(nextProps, nextState) {
+    return nextProps !== this.props || nextState !== this.state;
```
Oh I see, this is how you tell react to not re-render a component despite a changes taking place in the state/props.

Would implementing this for the peripheral components as well provide any speed boost?

Some quick searching also revealed the existence of [React.PureComponent](https://facebook.github.io/react/docs/react-api.html#react.purecomponent). Would it be possible to use that as the base class for all (or nearly all) components instead, to get a speed boost? This will make it so you don't have to manually write a `shouldComponentUpdate` for each component, but it does require purity.

kevinmasd replied:
I can try to check if this switch causes better improvements later, during state revision.

awli commented on dawn/fake-runtime/FakeRuntime.js:
```diff
@@ -33,11 +34,13 @@ let state = stateEnum.IDLE;
-  if (data.student_code_status !== 'IDLE') {
-    state = stateEnum.RUNNING;
-  } else {
+  // console.log(data);
```
Probably want to drop the comment if you're not going to log this at all.

awli commented on dawn/fake-runtime/FakeRuntime.js:
```diff
@@ -33,11 +34,13 @@ let state = stateEnum.IDLE;
+  } else if (data.student_code_status === 'ESTOP') {
+    state = stateEnum.ESTOP;
+  } else {
+    state = stateEnum.RUNNING;
```
consider a switch statement? Your preference though.

awli commented on dawn/main/ansible/Ansible.js:
```diff
@@ -1,35 +1,103 @@
+let runtimeSocket = null;
+let received = false;
+
+if (!Date.now) {
```
Are you ever running in an environment where you don't have Date.now?

awli commented on dawn/main/ansible/Ansible.js:
```diff
@@ -1,35 +1,103 @@
-  const status = data.studentCodeStatus ?
-    StudentCodeStatus.TELEOP : StudentCodeStatus.IDLE;
+  let status = null;
+  if (data.studentCodeStatus === 1) {
```
Magic numbers...

awli commented on dawn/renderer/components/Editor.js:
```diff
@@ -182,6 +187,19 @@ class Editor extends React.Component {
       throw err;
     });
+    ipcRenderer.send('NOTIFY_UPLOAD');
+    while (this.notificationHold === 1) {
```
Is this intentionally blocking the thread until notificationHold is not 1? Consider using a Promise or something else instead.

kevinmasd replied:
I'll look into Promises later; blocking is intentional.

kevinmasd committed:
[Dawn] Enumerables in Protocol

nikitakit commented:
Please make sure to open a GitHub issue about the busy-waiting loop before merging this. (Or even better, fix the busy-waiting. But I understand that getting a driving robot takes priority)

kevinmasd committed:
[Dawn] PER_MESSAGE wasn't correctly receiving values

kevinmasd committed:
[Dawn] Logging on Networking

kevinmasd committed:
[Dawn] Default IDLE if timeout detected

kevinmasd committed:
Merge branch 'master' of https://github.com/pioneers/PieCentral into dawn/protocolChanges

kevinmasd committed:
[Dawn] Peripherals updated to match protos

yizow commented on dawn/fake-runtime/FakeRuntime.js:
```diff
@@ -33,11 +34,15 @@ let state = stateEnum.IDLE;
+    case 'ESTOP':
+      state = stateEnum.ESTOP;
+      break;
+    default:
```
I think another case is 'auto' for autonomous code. look at the dawn/runtime spec for details.

kevinmasd committed:
[Dawn] Promising instead of Busy-Waiting

kevinmasd committed:
[Dawn] Lint Fixes

yizow committed:
[DAWN] TCP confirm refactor (#171)

Removes the busy wait while waiting for Runtime confirmation of student
code upload. Replaced with the usage of setTimeout.

kevinmasd committed:
[Dawn] Initial Try to handle EADDRUSE issues

kevinmasd committed:
[Dawn] Fake Runtime updated with new proto, Ansible code cleanup

kevinmasd committed:
[Dawn] Add Update.sh call, clean-up code

kevinmasd committed:
[Dawn] Cleaning Info Store

kevinmasd committed:
[Dawn] Fix Console Test

kevinmasd committed:
[Dawn] Set update connection to kill only on success of message

yizow commented on dawn/renderer/components/Editor.js:
```diff
@@ -136,93 +139,105 @@ class Editor extends React.Component {
+    ipcRenderer.send('NOTIFY_UPLOAD');
+    const waiting = () => {
+      let count = 0;
+      const waitPromise = (resolve, reject) => {
```
Move this to a saga

yizow commented on dawn/renderer/components/GamepadItem.js:
```diff
@@ -127,7 +127,7 @@ class GamepadItem extends React.Component {
 GamepadItem.propTypes = {
   gamepad: React.PropTypes.object,
-  index: React.PropTypes.number,
+  index: React.PropTypes.string,
```
revert and fix elsewhere.

yizow commented on dawn/renderer/components/UpdateBox.js:
```diff
@@ -51,28 +51,47 @@ class UpdateBox extends React.Component {
                 message: 'Update File Upload Failed.',
               });
               throw err2;
+            } else {
```
create dispatch and move this to saga.

yizow commented on dawn/renderer/reducers/info.js:
```diff
@@ -1,27 +1,36 @@
         ...state,
         connectionStatus: true,
+        robotState: action.robotState,
+        isRunningCode: (action.robotState === 1 ||
```
make enum

yizow commented on dawn/renderer/reducers/peripherals.js:
```diff
@@ -1,16 +1,35 @@
-    default:
+    }
+    case 'PERIPHERAL_RENAME': {
+      nextState[action.id] = state[action.id];
```
nop

yizow committed:
[DAWN] Fix gamepad index type bug

The index was being interpreted as a string, causing a runtime warning.
This change fixes that bug.

yizow committed:
[DAWN] Uses enum for checking runtime state

Adds a new enum runtimeStateEnum. Will need to be deprecated when
dawn and runtime protobufs get synced up

kevinji committed:
Merge branch 'master' into dawn/protocolChanges

kevinji committed:
[DAWN] eslint: Re-enable disabled rules

Instead, let's disable them on a case-by-case basis.

yizow committed:
[DAWN] Removed nop

kevinji committed:
[DAWN] eslint: Fix lint issues

kevinji committed:
[DAWN] package.json: No more react-devtools
